Fails to compile on Rust older than 1.45.0

Right now we're using the `paste` crate to have better control over identifiers in macros, but some of the macros are written in a way that's not compatible with rust < 1.45.0:

```
% rustup override set 1.42.0
info: using existing install for '1.42.0-x86_64-unknown-linux-gnu'
info: override toolchain for '…' set to '1.42.0-x86_64-unknown-linux-gnu'

  1.42.0-x86_64-unknown-linux-gnu unchanged - rustc 1.42.0 (b8cedc004 2020-03-09)

% cargo check
   Compiling autocfg v1.0.1
   Compiling memchr v2.3.4
   Compiling lazy_static v1.4.0
   Compiling semver-parser v0.7.0
   Compiling regex-syntax v0.6.21
    Checking bit_field v0.10.1
   Compiling bl602-pac v0.1.0 (/…/bl602-pac)
    Checking nb v1.0.0
    Checking vcell v0.1.3
    Checking bare-metal v1.0.0
   Compiling paste v1.0.4
   Compiling thread_local v1.0.1
   Compiling semver v0.9.0
   Compiling num-traits v0.2.14
   Compiling num-integer v0.1.44
   Compiling num-rational v0.3.2
   Compiling num-iter v0.1.42
    Checking embedded-hal v1.0.0-alpha.4 (https://github.com/rust-embedded/embedded-hal#eae6c995)
   Compiling rustc_version v0.2.3
   Compiling bare-metal v0.2.5
   Compiling aho-corasick v0.7.15
   Compiling regex v1.4.2
    Checking num-complex v0.3.1
   Compiling riscv-target v0.1.2
   Compiling riscv v0.6.0
    Checking num v0.3.1
    Checking embedded-time v0.10.1 (https://github.com/FluenTech/embedded-time#12e78c34)
    Checking bl602-hal v0.1.0 (/…/bl602-hal)
error[E0658]: procedural macros cannot be expanded to statements
   --> src/gpio.rs:89:13
    |
89  | /             paste::paste! {
90  | |             let glb = unsafe { &*pac::GLB::ptr() };
91  | |             glb.uart_sig_sel_0.modify(|_r, w| unsafe { w
92  | |                 .[<uart_ $sigi _sel>]().bits(mode)
93  | |             });
94  | |             }
    | |_____________^
...
100 |       impl_uart_sig!(UartSig0, "Uart signal 0 (type state)", sig_0, UartMux0, "Uart multiplexer peripherals for signal 0");
    |       --------------------------------------------------------------------------------------------------------------------- in this macro invocation
    |
    = note: for more information, see https://github.com/rust-lang/rust/issues/54727

error[E0658]: procedural macros cannot be expanded to statements
   --> src/gpio.rs:89:13
    |
89  | /             paste::paste! {
90  | |             let glb = unsafe { &*pac::GLB::ptr() };
91  | |             glb.uart_sig_sel_0.modify(|_r, w| unsafe { w
92  | |                 .[<uart_ $sigi _sel>]().bits(mode)
93  | |             });
94  | |             }
    | |_____________^
...
101 |       impl_uart_sig!(UartSig1, "Uart signal 1 (type state)", sig_1, UartMux1, "Uart multiplexer peripherals for signal 1");
    |       --------------------------------------------------------------------------------------------------------------------- in this macro invocation
    |
    = note: for more information, see https://github.com/rust-lang/rust/issues/54727

error[E0658]: procedural macros cannot be expanded to statements
   --> src/gpio.rs:89:13
    |
89  | /             paste::paste! {
90  | |             let glb = unsafe { &*pac::GLB::ptr() };
91  | |             glb.uart_sig_sel_0.modify(|_r, w| unsafe { w
92  | |                 .[<uart_ $sigi _sel>]().bits(mode)
93  | |             });
94  | |             }
    | |_____________^
...
102 |       impl_uart_sig!(UartSig2, "Uart signal 2 (type state)", sig_2, UartMux2, "Uart multiplexer peripherals for signal 2");
    |       --------------------------------------------------------------------------------------------------------------------- in this macro invocation
    |
    = note: for more information, see https://github.com/rust-lang/rust/issues/54727

… and more…
```

I assume this is because Rust v1.45.0 [stabilized function-like procedural macros](https://blog.rust-lang.org/2020/07/16/Rust-1.45.0.html#stabilizing-function-like-procedural-macros-in-expressions-patterns-and-statements)

We could update the MSRV, but I don't think it's necessary.

[full build.log](https://github.com/sipeed/bl602-hal/files/5764171/full.build.log)

--- a/src/gpio.rs
+++ b/src/gpio.rs
@@ -84,15 +84,15 @@ macro_rules! impl_uart_sig {
         pub fn into_uart1_rx(self) -> $UartMuxi<Uart1Rx> {
             self.into_uart_mode(7)
         }
+        paste::paste! {
         #[inline]
         fn into_uart_mode<T>(self, mode: u8) -> $UartMuxi<T> {
-            paste::paste! {
             let glb = unsafe { &*pac::GLB::ptr() };
             glb.uart_sig_sel_0.modify(|_r, w| unsafe { w
                 .[<uart_ $sigi _sel>]().bits(mode)
             });
-            }
             $UartMuxi { _mode: PhantomData }
+        }
         }
     }
     };
@@ -243,9 +243,10 @@ $(
         pub fn into_pull_down_input(self) -> $Pini<Input<PullDown>> {
             self.into_pin_with_mode(11, false, true, true)
         }
-        #[inline] fn into_pin_with_mode<T>(&self, mode: u8, pu: bool, pd: bool, ie: bool) -> $Pini<T> {
-            let glb = unsafe { &*pac::GLB::ptr() }; 
         paste::paste! {
+        #[inline]
+        fn into_pin_with_mode<T>(&self, mode: u8, pu: bool, pd: bool, ie: bool) -> $Pini<T> {
+            let glb = unsafe { &*pac::GLB::ptr() }; 
             glb.$gpio_cfgctli.modify(|_r, w| unsafe { w
                 .[<reg_ $gpio_i _func_sel>]().bits(mode) 
                 .[<reg_ $gpio_i _ie>]().bit(ie) // output
@@ -256,25 +257,23 @@ $(
             });
             // If we're an input clear the Output Enable bit as well, else set it.
             glb.gpio_cfgctl34.modify(|_, w| w.[<reg_ $gpio_i _oe>]().bit(!ie));
-        }
             $Pini { _mode: PhantomData }
+        }
         }
     }
 
     impl<MODE> $Pini<Input<MODE>> {
+        paste::paste! {
         /// Enable smitter GPIO input filter
         pub fn enable_smitter(&mut self) {
             let glb = unsafe { &*pac::GLB::ptr() };
-            paste::paste! {
-                glb.$gpio_cfgctli.modify(|_, w| w.[<reg_ $gpio_i _smt>]().set_bit());
-            }
+            glb.$gpio_cfgctli.modify(|_, w| w.[<reg_ $gpio_i _smt>]().set_bit());
         }
         /// Enable smitter GPIO output filter
         pub fn disable_smitter(&mut self) {
             let glb = unsafe { &*pac::GLB::ptr() };
-            paste::paste! {
-                glb.$gpio_cfgctli.modify(|_, w| w.[<reg_ $gpio_i _smt>]().clear_bit());
-            }
+            glb.$gpio_cfgctli.modify(|_, w| w.[<reg_ $gpio_i _smt>]().clear_bit());
+        }
         }
     }
 
@@ -332,54 +331,48 @@ $(
     impl<MODE> InputPin for $Pini<Input<MODE>> {
         type Error = Infallible;
 
+        paste::paste! {
         fn try_is_high(&self) -> Result<bool, Self::Error> {
             let glb = unsafe { &*pac::GLB::ptr() };
-            paste::paste! {
-                Ok(glb.gpio_cfgctl30.read().[<reg_ $gpio_i _i>]().bit_is_set())
-            }
+            Ok(glb.gpio_cfgctl30.read().[<reg_ $gpio_i _i>]().bit_is_set())
         }
 
         fn try_is_low(&self) -> Result<bool, Self::Error> {
             let glb = unsafe { &*pac::GLB::ptr() };
-            paste::paste! {
-                Ok(glb.gpio_cfgctl30.read().[<reg_ $gpio_i _i>]().bit_is_clear())
-            }
+            Ok(glb.gpio_cfgctl30.read().[<reg_ $gpio_i _i>]().bit_is_clear())
+        }
         }
     }
 
     impl<MODE> OutputPin for $Pini<Output<MODE>> {
         type Error = Infallible;
 
+        paste::paste! {
         fn try_set_high(&mut self) -> Result<(), Self::Error> {
             let glb = unsafe { &*pac::GLB::ptr() };
-            paste::paste! {
-                glb.gpio_cfgctl32.modify(|_, w| w.[<reg_ $gpio_i _o>]().set_bit());
-            }
+            glb.gpio_cfgctl32.modify(|_, w| w.[<reg_ $gpio_i _o>]().set_bit());
             Ok(())
         }
 
         fn try_set_low(&mut self) -> Result<(), Self::Error> {
             let glb = unsafe { &*pac::GLB::ptr() };
-            paste::paste! {
-                glb.gpio_cfgctl32.modify(|_, w| w.[<reg_ $gpio_i _o>]().clear_bit());
-            }
+            glb.gpio_cfgctl32.modify(|_, w| w.[<reg_ $gpio_i _o>]().clear_bit());
             Ok(())
+        }
         }
     }
 
     impl<MODE> StatefulOutputPin for $Pini<Output<MODE>> {
+        paste::paste! {
         fn try_is_set_high(&self) -> Result<bool, Self::Error> {
             let glb = unsafe { &*pac::GLB::ptr() };
-            paste::paste! {
-                Ok(glb.gpio_cfgctl32.read().[<reg_ $gpio_i _o>]().bit_is_set())
-            }
+            Ok(glb.gpio_cfgctl32.read().[<reg_ $gpio_i _o>]().bit_is_set())
         }
 
         fn try_is_set_low(&self) -> Result<bool, Self::Error> {
             let glb = unsafe { &*pac::GLB::ptr() };
-            paste::paste! {
-                Ok(glb.gpio_cfgctl32.read().[<reg_ $gpio_i _o>]().bit_is_clear())
-            }
+            Ok(glb.gpio_cfgctl32.read().[<reg_ $gpio_i _o>]().bit_is_clear())
+        }
         }
     }
 

